EDU-1867 (#2833)
temporalio/documentation: docs/evaluate/how-it-works.mdx

---
id: how-it-works
title: How Temporal Works
sidebar_label: How It Works
description: This guide provides a short overview of How Temporal Works.
toc_max_heading_level: 4
keywords:
  - temporal
  - evaluate-temporal
  - how-temporal-works

tags:
  - temporal
  - evaluate-temporal
  - how-temporal-works
---

> # How Temporal Works

Reviewer: Don't need this heading, as it duplicates the title.

> ---

Author: Thanks!

id: how-it-works
title: How Temporal Works
sidebar_label: How It Works
description: This guide provides a short overview of how Temporal works. 
toc_max_heading_level: 4
keywords:
  - temporal
  - evaluate-temporal
  - how-temporal-works
tags:
  - temporal
  - evaluate-temporal
  - how-temporal-works
---
Temporal offers an entirely new way to build scalable and reliable applications.

## Build Invincible Apps 

> In any complex system, failures are bound to happen. 

Author: I want to pull them in the first few sentences so I think I'll keep bound to happen for now.

> Temporal makes your applications reliable by default - ensuring that these failures don’t disrupt your application. 

Author: This is fantastic!

In a non-Temporal application, if a crash occurs, the progress of your application is lost. 

> The application has no memory of what happened before the failure, requiring extensive error-handling logic and messy recovery code to resume. 

Author: alternative: complex recovery code

This process is time-consuming and error-prone, making it hard to ensure reliability.

Temporal keeps track of the progress of your application. 

> If something goes wrong like a power outage, it helps your application pick up right where it left off - it’s like having the ultimate autosave. 

Author: Debating about the first suggestion. I think I like if something goes wrong for now.

By offloading the responsibility of failure management from the application to the platform, this removes the need for extensive recovery coding, testing, and maintenance tasks.

This shift simplifies the development process and if a failure or a crash happens, your business processes keep running seamlessly without interruptions.
This allows for developers to focus on business logic rather than infrastructure concerns and create applications that are inherently scalable and maintainable.

Thousands of developers trust Temporal for use cases like order processing, customer onboarding, and payment handling because it lets them build invincible applications that are resilient, durable, and _just work_. 
With Temporal, your applications keep running, no matter what.

## Temporal Application: The Building Blocks

> In a Temporal application, the code you write is made up of Workflows and Activities. 

Author: Yeah I'll have to think about this sentence.

Temporal isn’t a no-code Workflow engine - it is **Workflows-as-Code**.
Instead of dragging and dropping steps in a visual interface, you write your Workflows in code in your favorite programming language, code editor, and other tools. 
This gives you full control and flexibility over your business processes, allowing you to build exactly what you need. 

### Activities
Activities are the individual tasks in your Workflow - functions or methods depending on the programming language.
Activity tasks often interact with the outside world and are prone to failure such as sending emails, making network requests, writing to a database, or calling an API. 

Activities are called directly from within your Workflow code.

If an Activity fails, Temporal automatically retries it based on your configuration. 
Since Activities often rely on external systems, transient issues can occur - temporary but critical problems like network failures, timeouts, or service outages. 
You have full control over how often and how many times these retries should happen for each Activity. 

### SDK
Developers create Temporal applications by writing code, just like you would to create any other software. 

> To do this, you use an Temporal SDK, which stands for software development kit. 

Author: Vale instructs to use parentheses judiciously.

Temporal offers six SDKs: Java, Go, TypeScript, .NET, Python, and PHP. 
Installing the SDK allows you to easily add Temporal to your current projects without changing the tools you're already using. 
Temporal fits right into your existing tech stack.

The Temporal SDK also makes it easier to implement complex tasks such as sync/async execution, automatically retrying failed tasks, scheduling tasks to run regularly, managing complex workflows such as Saga compensation, and managing errors.

## Temporal Service 

Temporal has two main parts:
1. Your application (with Workflows and Activities)
2. The Temporal Service (a set of services and components)

At the heart of Temporal's architecture is the Temporal Service, which provides durability, scalability, and reliability for your application.
It communicates with the application, manages tasks, and maintains a detailed history of each event, which it durably persists to a database.
The Temporal Service manages critical tasks like making an API call or updating a database, by making sure they get executed, and then recording their completion.

One of the biggest advantages of the Temporal Service is how it handles failures. 

> The Temporal Service maintains a meticulous log of every step in your Workflows. 

Author: True. Good catch.

By keeping a history of every step in your Workflow, it ensures that even if something goes wrong — your Workflow can continue from the last successful point. 
The Temporal Service knows exactly where to resume without losing any work. 
This saves you from having to write complex error-handling code or painstaking recovery mechanisms yourself. 

You can run the Temporal Service on your own infrastructure or use Temporal Cloud, a managed service that handles operational overhead and offers scalability and expert support.

![Describing a Temporal Application and Temporal Service - How It Works](/img/how-it-works/htw_application_service_server.png)

## Workers
The real strength of Temporal comes from the combination of your application and the Temporal Service. 

Workers execute the actual tasks defined in your Workflows. 
The Temporal Service tracks and coordinates these Workers. 

Whenever your application needs to perform a task, like sending a notification or calling an processing a payment, the Temporal Service steps in. 

> This process creates a clear, organized list of what needs to be done. 

Author: I don't think I've explained task queue yet to be able to state it here (i.e - progressive disclosure). It's hard to explain the Temporal Service without a ton of Temporal stated terms.

---
id: how-it-works
title: How Temporal Works
sidebar_label: How It Works
description: This guide provides a short overview of how Temporal works. 
toc_max_heading_level: 4
keywords:
  - temporal
  - evaluate-temporal
  - how-temporal-works
tags:
  - temporal
  - evaluate-temporal
  - how-temporal-works
---
Temporal offers an entirely new way to build scalable and reliable applications.

## Build Invincible Apps 

> Temporal ensures your applications are reliable by default, handling everything from minor network outages to significant issues like unexpected reboots or hardware failures.

Reviewer 2: "Temporal ensures your applications are reliable" <- I feel like this is kind of vague still. If I was coming on this page, had no experience/knowledge of distributed systems, I would be confused by what it means by "makes your applications reliable and handles everything from network outages etc." I would be wondering:

- What makes an application 'reliable'?
- What do you mean by 'handles everything from minor network outages' etc.? 

I would flush out this paragraph a bit more personally!

- What does it mean that

Reviewer 3: I've asked @tomwheeler if he has a script available that you can steal from for his intro to Temporal for onboarding new employees.

By offloading the responsibility of failure management from the application to the platform, this removes the need for extensive coding, testing, and maintenance tasks.

This shift simplifies the development process and enables your application to seamlessly recover from disruptions.

> It does so by automatically reconstructing its previous state.

Reviewer 2: To a technical person, "reconstructing its previous state" might be confusing or unclear why that helps make an application more reliable. 

Maybe it would be helpful to use the analogy of comparing Temporal’s state recreation to a saved game in a video game that lets you pick up where you left off after a crash. Or a paused movie.
Example: "Temporal keeps track of everything your application does. So if something goes wrong, like a power outage, Temporal helps the application pick up right where it left off, just like resuming a video game right after a crash." 

I would also  emphasize the complexity and challenges involved in state management. For example, "Manual state management often requires careful handling of concurrency, ensuring data consistency, and managing failures during the save or reload process, all of which add significant overhead to the development process" because someone non-technical might wonder why saving the state is a complex thing to do

Author: Good points.

## Temporal Application: The Building Blocks

In a Temporal application, the code you write is made up of Workflows and Activities. 

### Workflow

Conceptually, a workflow is a sequence of steps. 
You've likely encountered workflows in your daily life, whether it's:
- Using a mobile app to transfer money
- Purchasing concert tickets
- Booking a vacation
- Ordering a meal delivery
- Filing an expense report
- Creating a new employee onboarding process

In Temporal, a Workflow defines the steps your application takes to complete a task, much like a blueprint for your application. 
It’s your business logic, written in code, outlining each step in your process. 
This could be moving money between bank accounts, processing orders, deploying cloud infrastructure, or training an AI model. 

Temporal isn’t a no-code workflow engine. 
It is **Workflows-as-Code**. 
Instead of dragging and dropping steps in a visual interface, you write your Workflows in code in your favorite programming language, code editor, and other tools. 
This gives you full control and flexibility over your business processes, allowing you to build exactly what you need. 

Workflows are resilient and stateful. 

> With Temporal, they are built to handle problems smoothly. 

Reviewer 2: This is vague. What do you mean by "handle problems smoothly"?

Author: Removed.

> They can pause, pick up where they left off, or try again automatically. 

Reviewer 2: Lines 51-53 are choppy because it's just one sentence after another. Can we connect them a bit more?

> They can pause, pick up where they left off, or try again automatically. 

Reviewer 2: Line 53 and 25 are extremely similar.

Author: Yeah I think sometimes saying things in different ways helps with comprehension.

If a failure or a crash happens, this ensures that your business processes keep running without interruptions.

### Activities

> Activities are the individual tasks within your Workflow.

Reviewer 2: Can be misleading. How would an Activity be different from a Workflow Task then?

Author: Updated it, let me know any recommendations. I'll check how Tom describes Activities too.

They handle jobs that interact with the outside world, like sending an email, making a network request, writing to a database, or calling an API. 

> These tasks are where issues might occur, such as network failures or service errors. 

Reviewer 2: "issues might occur" is vague. Maybe transient issues. Because we can have application-level issues, like an invalid credit card number, but that isn't going to be useful to put in an Activity, because retrying that won't help.

Author: Yeah transient issues is a term that's common in the distributed systems world but I'm not sure if it's common for the wide range of developers. So, I'll have to explain it first. It might be helpful to do an introduction of the term then explanation here since I think we use it pretty often in our explanations.

Activities are pieces of code, either functions or methods depending on the programming language, that perform these tasks. 
Activities are invoked in Workflow code. 

If an Activity fails, Temporal automatically retries it based on your configuration. 
You have full control over how often and how many times these retries should happen. 

> You can also define specific timeouts for each Activity, triggering a retry if the task takes too long to complete.

Reviewer 2: Line 66 is basically the same as line 65. When you say, "You can also.." but say the same thing, it might be redundant. I would remove line 65 and just keep line 66.

Author: Combined them!

### SDKs
Developers create Temporal applications by writing code, just like you would to create any other software. 
To do this, you use an Temporal SDK, which stands for software development kit. 

The Temporal SDK is an open source library that developers install in their application to use Temporal. 
It provides everything needed to build Workflows, Activities, and various other features in a specific programming language.

Temporal currently offers six SDKs: Java, Go, TypeScript, .NET, Python, and PHP. 
Installing the SDK allows you to easily add Temporal to your current projects without having to change the tools you're already using. 
Temporal fits right into your existing tech stack.

> The Temporal SDK also makes it easier to implement complex tasks such as handling jobs either in order or at the same time (sync/async execution), automatically retrying failed tasks, scheduling tasks to run regularly, managing complex workflows (SAGA compensation), and managing errors (advanced error handling).

Reviewer 2: I don't think you need 'managing errors (advanced error handling)'. I think just 'managing errors' is fine.

Author: Yeah I believe I took this from a list somewhere and turned it into a sentence that's more easy to take in for this context. Agreed that it's basically the same thing.

The Temporal Service knows exactly where to resume without losing any work. 
This saves you from having to write complex error-handling code or recovery mechanisms yourself. 

You can run the Temporal Service on your own infrastructure or use Temporal Cloud, a managed service that handles operational overhead and offers scalability and expert support.

![Describing a Temporal Application and Temporal Service - How It Works](/img/how-it-works/htw_application_service_server.png)

## Workers
The real strength of Temporal comes from the combination of your application and the Temporal Service. 
A Worker is a component within your application and provided by the Temporal SDK. 
Workers execute the actual tasks defined in your Workflows. The Temporal Service tracks and coordinates these Workers. 

Whenever your application needs to perform a task, like sending a notification or calling an processing a payment, the Temporal Service steps in. 
This process creates a clear, organized list of what needs to be done. 
It hands off tasks to these Workers, making sure they’re done correctly and efficiently while it keeps track of their progress. 

> Workers which are part of your application then carry out the tasks and run your code. 

Reviewer 2: ```suggestion
Workers, which are part of your application, then carry out the tasks and run your code.
```

This collaboration is crucial for building reliable, scalable, and durable applications. 
You can run multiple Workers — often dozens, hundreds, or even thousands — to improve application performance and scalability. 

![Describing a Temporal Application and Temporal Service and Workers - How It Works](/img/how-it-works/htw_application_service_workers.png)

:::info 
For a deep dive into how Temporal Service works, visit the [Temporal Encyclopedia page](/clusters).
:::

> ## Recovering from Failure

Reviewer 2: I feel like this section should be moved up before you introduce the components of Temporal. This answers the question of 'why temporal' before getting into the nitty gritty.

Author: This is supposed to be more of a summary where all the Temporal terms are now introduced so it's explaining Temporal with our terminology that they should now know.

In any complex system, failures are bound to happen, but how they are handled makes all the difference. 
Temporal ensures that these failures don’t disrupt your application. 
This reliability is achieved through **Durable Execution**.

In a traditional non-Temporal application, if a crash occurs, the progress of your Workflow is often lost. 
The application has no memory of what happened before the failure, requiring extensive error-handling and recovery code to resume. 
This process is time-consuming and error-prone, making it hard to ensure reliability.

> With Temporal, if your Workflow crashes, it automatically recreates its pre-crash state and continues as if the failure never happened. 

Reviewer 2: You've mentioned the points in this paragraph several times already, so it's kind of repetitive. Maybe it would be better to have an 'Overview' section and have the items in this paragraph as bullet points if you want to re-emphasize these points.

Reviewer 3: https://www.notion.so/temporalio/Cribbed-from-Tom-2f7b42438c8549289ba6c0dc1edabd73

Grabbed some great words you can crib from @tomwheeler

---
id: how-it-works
title: How Temporal Works
sidebar_label: How It Works
description: This guide provides a short overview of How Temporal Works.
toc_max_heading_level: 4
keywords:
  - temporal
  - evaluate-temporal
  - how-temporal-works

> 

Author: Done! Any suggestions or ideas on the text or images/structure?

tags:
  - temporal
  - evaluate-temporal
  - how-temporal-works
---

Temporal provides a comprehensive platform for building distributed applications with a focus on durability and reliability. 

## Resilient applications by default

Temporal ensures your applications are reliable by default, handling everything from minor network outages to significant issues like unexpected reboots, or hardware failures.
By offloading the responsibility of failure management from the application to the platform, this removes the need for extensive coding, testing, and maintenance tasks. 

This shift simplifies the development process and enables your application to seamlessly recover from disruptions. 
It does so by automatically reconstructing its previous state. 
Your application logic can be recovered, replayed, or paused from any previous point in the application process.

> ## How It Works

Reviewer: Consider the title of the page is already "How it works" - do we need this header here as well?

![HIW 1](/img/how-it-works/hiw-yourapplicationclient.png)

### Workflow
A Temporal application is the code you write, comprised of Workflow and Activities. Workflows are defined as code: either a function or an object method, depending on the language. 
A Temporal Workflow defines your overall business logic. That business logic might involve moving money between bank accounts, processing orders, deploying cloud infrastructure, training an AI model, or something else entirely. 
The Worker runs the Workflow and Activity code. 

> ![HIW 2](/img/how-it-works/hiw-configuretimeout.png)

Reviewer: This text precedes an image with "Your application (Worker)" -

But the header here is "Workflow" - so it doesn't seem to align.

Keep a close eye on the text in the images - and try to build more off of that.

---
id: how-it-works
title: How Temporal Works
sidebar_label: How It Works

> description: This guide provides a short overview of How Temporal Works.

Reviewer 2: Why random capitalization out of curiosity? How about: "This guide provides a short overview of how Temporal works."?

toc_max_heading_level: 4
keywords:
  - temporal
  - evaluate-temporal
  - how-temporal-works
tags:
  - temporal
  - evaluate-temporal
  - how-temporal-works
---

> Temporal offers an entirely new way to build scalable and reliable applications.

Reviewer 2: by...? Do we want to give a brief explanation?

Author: I think I want to keep it short and sweet and hook em in! 🎣

## Build Invincible Apps 

Temporal ensures your applications are reliable by default, handling everything from minor network outages to significant issues like unexpected reboots or hardware failures.
By offloading the responsibility of failure management from the application to the platform, this removes the need for extensive coding, testing, and maintenance tasks.

This shift simplifies the development process and enables your application to seamlessly recover from disruptions.

> It does so by automatically reconstructing its previous state.

Author: I want to try to keep short sentences.

> Your application logic can be recovered, replayed, or paused from any previous point in the application process.

Reviewer 2: Your application logic can be recovered, replayed, or paused from any previous point in your workflow, ensuring continuity even after failures.

Author: Added a sentence!

![Describing a Temporal Application - How It Works](/img/how-it-works/hiw-yourapplicationclient.png)

### Temporal Application

> A Temporal application is the code you write, comprised of Workflow and Activities. 

Reviewer 2: ```suggestion
A Temporal application is the code you write, composed of Workflow and Activities.
```

I changed it because "comprised of" sounded off to me, and then I googled it and found:

_Comprised of is an expression in English that means "composed of".[[1]](https://en.wikipedia.org/wiki/Comprised_of#cite_note-:0-1) This is thought by [language purists](https://en.wikipedia.org/wiki/Linguistic_purism) to be improper because to "comprise" (without the "of") can already mean to "be composed of"_

> A Temporal Workflow defines your overall business logic. That business logic might involve moving money between bank accounts, processing orders, deploying cloud infrastructure, training an AI model, or something else entirely.

Reviewer 2: ```suggestion
A Temporal Workflow defines your overall business logic. That business logic might involve moving money between bank accounts, processing orders, deploying cloud infrastructure, or training an AI model.
```

> A Temporal Workflow defines your overall business logic. That business logic might involve moving money between bank accounts, processing orders, deploying cloud infrastructure, training an AI model, or something else entirely.

Reviewer 2: Would add something about the resiliency of Workflows.

Maybe: "Workflows are resilient and can be paused, resumed, or retried automatically, ensuring that your business processes are fault-tolerant."

Author: I changed up the structure to where those points are mentioned a lot.

The Worker runs the Workflow and Activity code.

![Workflow Function Execution - How It Works](/img/how-it-works/hiw-workflowexecution.png)

> ### Activity

Reviewer 2: I think different titles might be better. Even something just like, "What is an Activity?"

Author: Hmm yeah maybe in another iteration we can play around with the titles. I changed it to question marks as a test but it looked a bit weird with the different levels of headings.

> Activities represent code (either a functions or method) that handle tasks prone to errors or failure, such as network requests, calling a service, sending an email, or generating random numbers.

Reviewer 2: ```suggestion
Activities represent code (either a function or a method) that handles tasks prone to errors or failure, such as network requests, calling a service, sending an email, or generating random numbers.
```

Reviewer 2: Well, generating random numbers isn't something that's prone to errors or failures, but it's in non-deterministic. Generating random numbers is something put in an Activity, because we want to encapsulate non-deterministic code in Activities, but the way it's written here, it's as if generating random numbers is prone to errors or failures. How about something like:

Activities represent code (either a function or a method) that handles tasks prone to errors or failure, such as network requests, calling a service or sending an email. Activities also encapsulate non-deterministic code such as a function which generates random numbers.

Reviewer 2: Well, generating random numbers isn't something that's prone to errors or failures, but it's in non-deterministic. Generating random numbers is something put in an Activity, because we want to encapsulate non-deterministic code in Activities, but the way it's written here, it's as if generating random numbers is prone to errors or failures. How about something like:

Activities represent code (either a function or a method) that handles tasks prone to errors or failure, such as network requests, calling a service or sending an email. Activities also encapsulate non-deterministic code such as a function which generates random numbers.

Reviewer 2: Also, kinda being nit-picky about "Activities represent code" part. Activities ARE code. How about:

Activities are a function or a method that handles tasks prone to errors or failure.....

> Activities represent code (either a functions or method) that handle tasks prone to errors or failure, such as network requests, calling a service, sending an email, or generating random numbers.

Reviewer 2: I think that we should explain why code prone to failure/non-deterministic code is encapsulated in Activities. Reader might be confused why that is. The reason is because we can retry them over and over, making Temporal resilient against transient issues. I think that's super important to mention the retry policy.

> Activities represent code (either a functions or method) that handle tasks prone to errors or failure, such as network requests, calling a service, sending an email, or generating random numbers.

Reviewer 2: Also, kinda being nit-picky about "Activities represent code" part. Activities ARE code. How about:

Activities are a function or a method that handles tasks prone to errors or failure.....

> Activities are invoked in Workflow code and the Temporal Service coordinates with the application to execute them.

Reviewer 2: Do we want to mention that Activities have the retry capability? I think that's super important.

![Continuing a Workflow - How It Works](/img/how-it-works/hiw-continuingtheworkflow.png)

### Temporal Service

Temporal consists of two components, the [Temporal Service](/evaluate/major-components#temporal-service) and an SDK in your preferred language such as Go, .NET, Java, Python, PHP, or Typescript.

> The Temporal Service logs the progress and state of your application.

Reviewer 2: "logs the progress and state" is kind of off to me in a Temporal Context. How about: "The Temporal Service records every event in the Workflow’s lifecycle, so that the Workflow can continue from the last successful state in the case of a failure" or something like that

Author: I edited this after receiving this feedback so these points are mentioned now.

Once the application finishes executing the Activity, it communicates with the Temporal Service, which records the Activity's completion in the Workflow's Event History.

> The Activity invocation is handled by the Temporal Service.

Reviewer 2: Kind of misleading I think, the Temporal Service manages the scheduling of the Activity, but the Worker actually executes the Activity.

Maybe change to: "The Temporal Service schedules the Activity and creates a Task, which is picked up and executed by a Worker."

The Temporal Service updates the Event History and creates a Task for executing the Activity.

![Activity Execution - How It Works](/img/how-it-works/hiw-activityexecution.png)

### Recovering from Failure

> A Temporal Workflow is basically a process or sequence of events you want to make sure will execute reliably.

Reviewer 2: Consider rewording? The "basically" is a bit informal IMO. How about: "A Temporal Workflow is a sequence of events that are designed to execute reliably, even in the face of failures."

They can run—and keep running—for years, even if the underlying infrastructure fails.
If the execution of your Workflow code crashes, Temporal automatically recreates its pre-crash state, allowing the Workflow to continue execution as if it never even happened.
Temporal provides built-in solutions for timers, event sourcing, state checkpointing, retries, and timeouts, removing the necessity for custom implementations.

![Recovering from Failure - How It Works](/img/how-it-works/hiw-recoveringfromfailure.png)

### Concluding the Workflow

In this example, there are two Activities. The second Activity is complete, but the Workflow function still has additional code to run.
The Temporal Service creates a new Task requesting that the application run the additional code.

![Concluding the Workflow - How It Works](/img/how-it-works/hiw-concludingtheworkflow.png)

## Common Misconceptions

> The Temporal Service does not execute your code!

Reviewer 2: Hmm, I personally think the tone can be softened :) Without the !

How about: "It's important to note that the Temporal Service does not execute your code."

> Your application code, including Workers, Workflows, and Activities, lives on your computers. 

Reviewer 2: Not always locally. How about: "Your application code, including Workers, Workflows, and Activities, runs on your own infrastructure, whether that's on local machines, servers, or cloud environments."
